src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/WorkspaceFactory.cs

```diff
@@ -146,10 +148,20 @@ var buildTransformBlock
                 cancellationToken: cancellationToken),
 
             buildTransformBlock.LinkTo(orderingBlock, DataflowOption.PropagateCompletion),
-
-            ProjectDataSources.JoinUpstreamDataSources(joinableTaskFactory, _projectService.Services.FaultHandler, source.ActiveConfiguredProjectSource, source.ProjectBuildRuleSource)
         });
 
+        // During solution load we run evaluations but not DTB. CPS has a feature that detects when the main thread
+        // needs DTB results and allows the DTB to run before solution load completes, in order to avoid deadlocks.
+        // While it's fine to create a dataflow subscription for build data during solution load, the joining of
+        // upstream data sources can cause CPS to accidentally think that DTB is blocking the UI thread. There's only
+        // one gate for the whole solution, so even a single project in this state will cause all projects to build.
+        // In order to avoid this, we defer joining upstream sources until after the solution has loaded.
+        _ = _tasksService.SolutionLoadedInHost.ContinueWith(
+            _ => workspace.ChainDisposal(ProjectDataSources.JoinUpstreamDataSources(joinableTaskFactory, _projectService.Services.FaultHandler, source.ActiveConfiguredProjectSource, source.ProjectBuildRuleSource)),
+            CancellationToken.None,
```

Comment:
suggest to chain to the cancellationToken of the workspace object (i think there is a cancellationToken when it is disposed). It is just to ensure we abort this continuation, if loading is aborted earlier. (Not sure whether it can happen in real product)

Reply:
I wondered about that. There's an existing usage of `CancellationToken.None` in the fault handling case, and I thought it might be safer to avoid cancelling those operations. That said, if we're tearing down the project (which is the main cancellation here) then I don't think either matters.